### PR TITLE
Convert vmextensions service to SDKv2

### LINKED ISSUE #3443: Update VM extensions service to SDK v2

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Converts the vmextensions service to azure-sdk-for-go version 2 and the `asyncpoller` framework for long-running operations.

**Which issue(s) this PR fixes**:

Fixes #3443

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: azure/services/vmextensions/client.go [hunk 1]
```diff
@@ -18,138 +18,103 @@ package vmextensions
...
+func newClient(auth azure.Authorizer) (*azureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create virtualmachineextensions client options")
+	}
+	factory, err := armcompute.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armcompute client factory")
+	}
+	return &azureClient{factory.NewVirtualMachineExtensionsClient()}, nil
```
Comment (willie-yao) [inline azure/services/vmextensions/client.go:46]: nit: Is it better to place this logic into a separate helper function like in the  `natgateways` service to keep things consistent? https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/files#diff-0e0342ee34be437e8b2443243544f4d3a79fc4655bf9552bffba2f187f4d18ebR46

File: azure/services/vmextensions/client.go [hunk 2]
```diff
@@ -18,138 +18,103 @@ package vmextensions
...
 
-	vmextension, ok := parameters.(compute.VirtualMachineExtension)
+	vmextension, ok := parameters.(armcompute.VirtualMachineExtension)
 	if !ok {
-		return nil, nil, errors.Errorf("%T is not a compute.VirtualMachineExtension", parameters)
+		return nil, nil, errors.Errorf("%T is not an armcompute.VirtualMachineExtension", parameters)
 	}
 
-	createFuture, err := ac.vmextensions.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), vmextension)
+	opts := &armcompute.VirtualMachineExtensionsClientBeginCreateOrUpdateOptions{ResumeToken: resumeToken}
```
Comment (willie-yao) [inline azure/services/vmextensions/client.go:73]: Another consistency nit: there is a log message in a previously merged PR https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/files#diff-0e0342ee34be437e8b2443243544f4d3a79fc4655bf9552bffba2f187f4d18ebR83

File: azure/services/vmextensions/client.go [hunk 3]
```diff
@@ -18,138 +18,103 @@ package vmextensions
...
+// request to Azure and if accepted without error, the func will return a Poller which can be used to track the ongoing
 // progress of the operation.
-func (ac *azureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter) (future azureautorest.FutureAPI, err error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.AzureClient.DeleteAsync")
+func (ac *azureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter, resumeToken string) (poller *runtime.Poller[armcompute.VirtualMachineExtensionsClientDeleteResponse], err error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.azureClient.DeleteAsync")
 	defer done()
 
-	deleteFuture, err := ac.vmextensions.Delete(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName())
+	opts := &armcompute.VirtualMachineExtensionsClientBeginDeleteOptions{ResumeToken: resumeToken}
```
Comment (willie-yao) [inline azure/services/vmextensions/client.go:101]: Same comment as above ^^

**Review by willie-yao — COMMENTED** [review]
Just a few nits from my end!

File: azure/services/vmextensions/client.go (hunk 1 above)
Comment (mboersma) [reply azure/services/vmextensions/client.go:46]: I guess it's a style thing, but I didn't see value in having the `newClient` func separate from the `newNatGatewaysClient` func, so I've made all the new PRs use this style, and I cleaned `natgateways` up to match in #3887.

File: azure/services/vmextensions/client.go (hunk 2 above)
Comment (mboersma) [reply azure/services/vmextensions/client.go:73]: That log message was meant more for debugging (I realized after it merged), so I've kept it out of subsequent SDKv2 PRs like this one. I can remove it from natgateways too at some point.

File: azure/services/vmextensions/client.go (hunk 3 above)
Comment (mboersma) [reply azure/services/vmextensions/client.go:101]: Same answer as above 😄: I meant to keep the log line out, although that's not consistent with the SDKv2 change already merged.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**